pkg/controllers/teammembership/teammembership_updater_controller.go

```diff
@@ -0,0 +1,346 @@
+// SPDX-FileCopyrightText: 2024 SAP SE or an SAP affiliate company and Greenhouse contributors
+// SPDX-License-Identifier: Apache-2.0
+
+package teammembership
+
+import (
+	"context"
+	"errors"
+	"fmt"
+	"sync"
+	"time"
+
+	greenhousev1alpha1 "github.com/cloudoperators/greenhouse/pkg/apis/greenhouse/v1alpha1"
+	"github.com/cloudoperators/greenhouse/pkg/scim"
+	apierrors "k8s.io/apimachinery/pkg/api/errors"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
+	"k8s.io/client-go/tools/record"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/builder"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+	"sigs.k8s.io/controller-runtime/pkg/event"
+	"sigs.k8s.io/controller-runtime/pkg/handler"
+	"sigs.k8s.io/controller-runtime/pkg/log"
+	"sigs.k8s.io/controller-runtime/pkg/predicate"
+	"sigs.k8s.io/controller-runtime/pkg/reconcile"
+)
+
+type TeamMembershipUpdaterController struct {
+	client.Client
+	recorder          record.EventRecorder
+	ScimBaseURL       string
+	ScimBasicAuthUser string
+	ScimBasicAuthPw   string
+	scimClient        *scim.ScimClient
+	teamsByName       map[string]greenhousev1alpha1.Team
+}
+
+//+kubebuilder:rbac:groups=greenhouse.sap,resources=teammemberships,verbs=get;list;watch;create;update;patch
+//+kubebuilder:rbac:groups=greenhouse.sap,resources=teammemberships/status,verbs=get;update;patch
+//+kubebuilder:rbac:groups=greenhouse.sap,resources=teammemberships/finalizers,verbs=update
+//+kubebuilder:rbac:groups=greenhouse.sap,resources=organizations,verbs=create;update;patch;delete
+//+kubebuilder:rbac:groups=greenhouse.sap,resources=teams,verbs=create;update;patch;delete
+//+kubebuilder:rbac:groups="",resources=events,verbs=get;list;watch;update;patch
+
+// SetupWithManager sets up the controller with the Manager.
+func (r *TeamMembershipUpdaterController) SetupWithManager(name string, mgr ctrl.Manager) error {
+	r.Client = mgr.GetClient()
+	r.recorder = mgr.GetEventRecorderFor(name)
+
+	if r.ScimBaseURL == "" {
+		return errors.New("scim base URL required but not provided")
+	}
+	if r.ScimBasicAuthUser == "" {
+		return errors.New("scim basic auth user required but not provided")
+	}
+	if r.ScimBasicAuthPw == "" {
+		return errors.New("scim basic auth pw required but not provided")
+	}
+
+	scimConfig := scim.Config{
+		RawURL:   r.ScimBaseURL,
+		AuthType: scim.Basic,
+		BasicAuthConfig: &scim.BasicAuthConfig{
+			BasicAuthUser: r.ScimBasicAuthUser,
+			BasicAuthPw:   r.ScimBasicAuthPw,
+		},
+	}
+	scimClient, err := scim.NewScimClient(scimConfig)
+	if err != nil {
+		return err
+	}
+	r.scimClient = scimClient
+
+	r.teamsByName = make(map[string]greenhousev1alpha1.Team)
+
+	return ctrl.NewControllerManagedBy(mgr).
+		Named(name).
+		For(&greenhousev1alpha1.TeamMembership{}).
```

Comment:
This controller should watch the `Team` object, as this object carries the MappedIDPGroup necessary for getting the members of a team from the SCIM API. These members should then be persisted into the respective (a.k.a. equally named) `TeamMembership` resource.

I would go as far as we do not need to watch other resources (`TeamMemberships` or `Organizations`) as the actual changes that we want to know happen in SCIM and we cannot reconcile on those. That is why the timely requeue suffices. WDYT @IvoGoman

Reply:
I think the controller should be for the `TeamMembership` and is triggered either periodically to fetch changed from SCIM or if the Team or Org has changed.
Team and Org bring in the IDP Group which is used by the TeamMembership. So in case these change we need to know.
Also the Watch on the Org allows us to create an artificial reconcile event that will create the `<org>-admin` TeamMembership.

Reply:
@uwe-mayer and I discussed this and come up with this:

- Controller is `For(Team).Owns(TeamMembership)`
- Do not bother about creating the Org Admin Team in this controller. This controller should expect the teams to be the single source of truth